docs/denotational-design.md

```diff
@@ -1,214 +1,228 @@
+-- are equated and found is opaque, but we can think of them as
+-- branch or tag labels.
+getHistory :: Eq a => History a -> Repo a -> Maybe (History a)
+μ getHistory history repo =
+  find history (μ repo)
+
+-- Find if a particular artefact occurs in 0 or more histories.
+findInHistories :: a -> [History a] -> [History a]
+μ findInHistories a histories =
+  filterMaybe (findInHistory a) histories
+
+-- Find a particular artefact is in a history.
+findInHistory :: Eq a => a -> History a -> Maybe a
+μ findInHistory a history = find (== a) (μ history)
+
+-- A special Label that guarantees a starting point, i.e. ~
+root :: Label
+
+-- Get the difference between two directory views.
+diff :: Directory -> Directory -> Diff
+
+-- List the current file or directories in a given Directory view.
+listDirectory :: Directory -> NonEmpty (Label, SystemType)
+μ listDirectory directory = map f $ snd (μ directory)
+  where
+    f = \case
+      Left dir -> (fst dir, IsDirectory)
+      Right file -> (fst file, IsFile)
+
+fileName :: File -> Label
+μ fileName file = fst (μ file)
 
--- Getting the [Directory] from CommitHistory should be equivalent to building
--- the [File] and getting their Directory
-files_and_directories :: Property
-files_and_directories = ∀ history. getHistoryDirectories history ≡ map fileDirectory (getFiles (getCommits history))
+findFile :: NonEmpty Label -> Directory -> Maybe File
+μ findFile (label :| labels) directory =
+  let (label, artefacts) = (μ directory)
+  if label == label' then go labels artefacts else Nothing
+  where
+    findFileWithLabel :: Foldable f => Label -> f (Either Directory File) -> Maybe File
+    findFileWithLabel label = find (\artefact -> case artefact of
+      Left _     -> False
+      Right file -> fileLabel == label)
+
+    go :: [Label] -> [Either Directory File] -> Just File
+    go [] _ = Nothing
+    go [label] directories = findMaybe (fileWithLabel label) directories
+    go (label:labels) directories = go labels $ find ((label ==) . fst) onlyDirectories directories
+
+onlyDirectories :: Foldable f => f (Either Directory File) -> [Directory]
+onlyDirectories = filter isLeft . toList
+
+getSubDirectories :: Directory -> [Directory]
+μ getSubDirectories directory = foldMap f $ snd (μ directory)
+  where
+    f :: Either Directory File -> [Directory]
+    f = either pure []
+
+-- Definition elided
+findDirectory :: NonEmpty Label -> Directory -> Maybe Directory
+
+-- Definition elided
+fuzzyFind :: Label -> [Directory]
+
+-- A Git Snapshot is grabbing the HEAD commit of your History
+-- and turning it into a Directory
+gitSnapshot :: Snapshot [Commit]
+μ gitSnapshot = getDirectoryPtr . head
+
+-- Opaque and defined by the backend
+getDirectoryPtr :: Commit -> Directory
+
+-- A Pijul history is semantically applying the patches in a
+-- topological order and achieving the Directory view.
+pijulHistory :: Snapshot Patch
+μ pijulHistory = foldl pijulMagic mempty
```

Comment:
@kim Actually this seems to be broken as well. We would actually need to do `foldl pijulMagic rootDir` where `rootDir :: Directory`, but that's not possible since what do we put in the `NonEmpty`? :thinking:

Reply:
Unless we have:
```haskell
rootDir :: Directory
rootDir = ("~", Right (".root", empty) :| [])
```
i.e. the root directory is the root label with a root file.
```

Reply:
It feels a bit hacky though... But we do need some idea of a `root` at the same time so that we can begin a traversal.

Reply:
```haskell
gitRoot = ("~", Left (".git", ...) :| [])
pijulRoot = ("~", Left (".pijul", ...) :| [])
```

is accurately describing reality ;)

Except that those directories are magical and should not be browseable -- does that need to be captured in the semantics?

Reply:
But what's the first file we tie into those? :thinking: Or maybe this is some magic opaque trait that is `getRoot :: Directory`.

Reply:
How about something like this:

```haskell
type DirectoryContents
μ DirectoryContents = IsRepo | Directory | File

-- Opaque representation of repository state directories (e.g. `.git`, `.pijul`)
-- Those are not browseable, but have to be present at the repo root 'Directory'.
type IsRepo

type Directory
μ Directory = (Label, NonEmpty DirectoryContents)

emptyRepoRoot :: Directory
μ emptyRepoRoot = ("~", IsRepo :| [])
```

Reply:
Oh that's cool! Because we can have sub vcs directories inside directories!

Reply:
One downside of this is that it's possible to represent:
```haskell
wut :: Directory
wut = ("wut", IsRepo :| [IsRepo, IsRepo, IsRepo, IsRepo])
```
But we can hide those details from the user I guess :thinking:

Reply:
Hm. I'm sure this can be fixed somehow using `absurd` 🤣

Reply:
Or just regular smart constructors :eyes: :sweat_smile: Going forward with this design :+1: